accounts/manager.go

```diff
@@ -34,8 +35,34 @@ type Manager struct {
 
 	feed event.Feed // Wallet feed notifying of arrivals/departures
 
-	quit chan chan error
-	lock sync.RWMutex
+	quit      chan chan error
+	lock      sync.RWMutex
+	nonceLock accountNonceLocker // Used to lock account nonces, to prohibit nonce reuse
+}
+
+type accountNonceLocker struct {
+	mut   sync.Mutex
+	locks map[common.Address]*sync.Mutex
+}
+
+// LockNonce Locks a particular nonce for an account, to prevent another tx getting the same nonce until this lock has been released
+// This should be used when a nonce has not been provided by the user,
+// and we get a nonce from the pools. The mutex prevents the (an identical nonce) from being
+// read again during the time that the first transaction is being signed.
+func (am *Manager) LockNonce(address common.Address) {
+	if _, exist := am.nonceLock.locks[address]; !exist {
+		am.nonceLock.mut.Lock()
+		if _, exist := am.nonceLock.locks[address]; !exist {
```

Comment:
You're kind of going against Go's principles here though and I'm not sure whether that's a good idea.

`defer` is executed at the end of the function's scope, including `return` statement but excluding the actual returning (i.e. setting of return variables, both named and unnamed) and the nicety is that it will be executed regardless of where the `return` is.

The dual-check is really, really odd and there's an argument for infinite amount of these checks. I strongly suggest this is written as:


```go
func (am *Manager) LockNonce(address common.Address) {
	am.nonceLock.mut.Lock()
        defer am.nonceLock.mut.Unlock()

	if _, exist := am.nonceLock.locks[address]; !exist {
		am.nonceLock.locks[address] = new(sync.Mutex)
	}
	am.nonceLock.locks[address].Lock()
}
```

IMO this is better code and less cause for confusion.